---
title: "Introduction to ebGSEA"
author:
- name: "Andrew E. Teschendorff"
  affiliation: 
  - CAS Key Lab of Computational Biology, PICB, SINH
- name: "Tianyu Zhu"
  affiliation:
  - CAS Key Lab of Computational Biology, PICB, SINH
date: "`r Sys.Date()`"
package: ebGSEA
output:
  BiocStyle::html_document:
    theme: readable
bibliography: ebGSEA.bib
vignette: >
  %\VignetteIndexEntry{Empirical Bayes Gene Set Enrichment Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r vignette-options, echo=FALSE, warning=FALSE}
require(BiocStyle)
```

# Introduction

Gene Set Enrichment Analysis is one of the most common tasks in the analysis of omic data, and is critical for biological interpretation. In the context of Epigenome Wide Association Studies (EWAS), which typically rank individual cytosines according to the level of differential methylation, enrichment analysis of biological pathways is challenging due to differences in CpG/probe density between genes. Here we use an empirical Bayes Gene Set Enrichment Analysis (ebGSEA) algorithm, which does not rank CpGs but genes according to the overall level of differential methylation of its CpGs/probes, allowing unbiased and sensitive detection of enriched pathways. ebGSEA can be applied to any EWAS that uses Illumina HM450k and EPIC beadarrays. For details please refer to our publication listed at the end of this tutorial[@dong2019ebgsea].

# Tutorial Example

## Differentially methylated genes in a buccal swab EWAS of smoking

To illustrate the functions, we use a subset of a HM450k buccal swab dataset[@teschendorff2015correlation], containing 325 buccal swab samples with smoking-pack-years as phenotype, and 7933 CpG probes. The beta value matrix has been processed, with samples by column and probes by row.

```{r}
load("ebgseaDATA.rda");
ls();
dim(dataSMK.m)
```

To rank genes according to the levels of differential methylation of probes mapping to each gene, we apply the global test[@goeman2004a] function. Here are the inputs for the `doGT` function:

* pheno.v: A vector containing the phenotype information, matched to the columns of the input DNAm data matrix. In our case this contains the smoking pack years information.
* data.m: The matrix of DNAm beta values with probes along rows and samples along columns. Missing values are not allowed and should be imputed beforehand.
* model: The regression model for the global test. Default is “linear”.
* array: Array type for the input data. “450k” for Illumina HumanMethylation450 data and “850k” for Illumina MethylationEPIC data.
* ncores: Number of cores used for parallel running. (default = 4)

To run `doGT` :
```{r}
library(ebGSEA)
sgt.m <-doGT(phenoSMK.v,dataSMK.m,array="450k",ncores=4)
head(sgt.m)
```

The output `sgt.m` is a matrix with rows labeling genes, ordered according to their overall level of differential methylation. The last column gives us the number of CpGs/probes mapping to the given gene.

## Pathway enrichment analysis with Wilcox Rank Sum Test and Known-Population Median Test

The next step is to apply the function `doGSEAwt` to do a pathway enrichment analysis in a threshold independent manner, using the Wilcox test (WT) and the Known-Population Median Test (KPMT). Here are the input parameters:

* rankEID.m: The matrix output object from `doGT` function, with rows labeling genes. Rownames of the matrix should be Gene Entrez IDs.
* ptw.ls: List of vectors consisting of Gene EntrezIDs of genes pathways of interest. For a pathway-database we use 8567 biological terms from the Molecular Signatures Database[@subramanian2005gene], by invoking `data("MSigDB-28Feb14-data")`.
* ncores: Number of cores used for parallel running. (default = 4)
* minN: For each pathway, the minium number of genes(i.e. available in the ranked gene list) to conduct GSEA. If less than this value, the P-value of this pathway would be set 1. (default = 5)
* adjPVth: Adjusted P-value threshold to declare a pathway to be significantly enriched. P-value was derived from Wilcoxon rank sum test and adjusted with BH method. (default = 0.05)

```{r warning=FALSE}
data("MSigDB-28Feb14-data")
topGSEA.lm <- doGSEAwt(rankEID.m = sgt.m, ptw.ls = listEZ.lv, ncores = 4, minN = 5,adjPVth = 0.05)
```
The output is a list of three objects: 

* Rank(P): A matrix showing enriched pathways ranked by adjusted Wilcox test P-values.
* Rank(AUC): A matrix showing enriched pathways ranked by AUC.
* Genestat: Lists of gene symbols in each enriched pathway. Each object contains the statistic and P-value from global test of each gene.

```{r}
topGSEA.lm[[1]]
```

The columns of this matrix indicate the following:

* nREP: Number of genes mapped in this pathway and present on array
* AUC: Area under curve from Wilcox test
* P(WT): P-value from Wilcox Test
* P(KPMT): P-value from Known Population Median Test
* adjP: Adjusted P-value for each pathway, using BH method

We can see that the top-ranked biological term is related to nasopharyngeal carcinoma, which is a sensible result because the nasopharynx is exposed to smoke carcinogens, and smoking is indeed a major risk factor for this type of cancer.

The third list Genestat consists of the statistic and P-value from global test, of each gene in a specific enriched pathway. To see the genes in the first pathway:
```{r}
head(topGSEA.lm$Genestat[[1]])
```

We can plot the AUC and adjP for each enriched pathway:
```{r}
plot(x = topGSEA.lm[[1]][,2], y = -log10(topGSEA.lm[[1]][,5]), xlab ="AUC", ylab = "-log10(adjP)", main = "AUC and adjP for each enriched pathway", pch = 21, bg = "red");
```

## Pathway enrichment analysis with Fisher's Exact Test

Additionally, ebGSEA allows users to do GSEA in the more traditional way on a group of specified CpGs/Genes with Fisher's Exact Test, without the need of ranking genes.

If your input is a set of CpGs (say a list of top-ranked DMCs), you may use `selEIDfromSelCpG` function to first derive a corresponding list of genes. This function implements a binomial test to determine the statistical significance of a gene being selected on account of the number of CpGs mapping to the gene. Here are the input parameters:

* selCpG.v: A vector of user selected CpGs.
* allCpG.v: A vector representing all CpGs in the input DNAm data matrix.
* pvth: P-value threshold to infer the number of selected CpGs mapped to a gene is significant or not in a binomial test. (default = 0.3/length(selCpG.v))
* array: Array type for the input CpGs. "450k" for Illumina HumanMethylation450 data and "850k" for Illumina MethylationEPIC data.

To illustrate this function, we import a set of selected CpGs derived from a more complete buccal swab dataset (i.e. over 400k CpGs) by invoking `data("sampleCpG")`. This set of 40626 CpGs exhibit differential methylation associated with smoking pack-years.

```{r}
data("SampleCpG")
sigEID.ls <- selEIDfromSelCpG(selCpG.v = sampleCpG.v, allCpG.v = allCpG.v, array = "450k")
```

The group of CpGs are significantly mapped to 255 genes. 
```{r}
summary(sigEID.ls)
```

Now we can apply the `doGSEAft` function to do a pathway enrichment analysis with Fisher's Exact Test. Here are the input parameters:

* selEID.v: A vector of selected Entrez Gene ID.
* ptw.ls: Lists of Gene EntrezID in each pathway of interest. You can get the 8567 biological terms from Molecular Signatures Database by `data("MSigDB-28Feb14-data")`.
* allEID.v: A vector of the universal set of Entrez Gene ID which you select genes from.
* ncores: Number of cores used for parallel running. (default = 4)
* minN: For each pathway, the minium number of genes(i.e. available in the ranked gene list) to conduct GSEA. If less than this value, the p value of this pathway would be set 1. (default = 5)
* adjPVth: Adjusted p value threshold to infer a pathway to be significantly enriched or not. P value was derived from Wilcoxon rank sum test and adjusted with BH method. (default = 0.05)

```{r message=FALSE}
topGSEAft.lm <- doGSEAft(selEID.v = sigEID.ls$selEID, ptw.ls = listEZ.lv, allEID.v = names(mapEIDto450k.lv), ncores = 1, adjPVth = 0.05)
```
The output of `topGSEAft.lm` consists of the following items:

* Rank(P): A matrix showing enriched pathways ranked by adjusted Fisher's Exact Test P-values. "nREP" is the number of genes in the pathway, "nOVL" is the number of selected genes in the pathway, "OR" is the odds ratio of Fisher's Exact Test, "P" is the P-value of Fisher's Exact Test, "adjP" is the adjusted P-value of Fisher's Exact Test (method='BH'), "Genes" is all the selected genes in the pathway.
* Rank(OR): A matrix showing enriched pathways ranked by odds ratio. The columns are samely defined as in Rank(P).

There are 79 enriched pathways identified by Fisher's exact test, each with 6 features:
```{r}
summary(topGSEAft.lm)
```

```{r}
head(topGSEAft.lm$`Rank(P)`)
```
We can see that a lung cancer biological term appears at the top, which again is meaningful since lung cancer is strongly related to smoking.

Finally, we can plot the odds ratio and adjPval for each enriched pathway.
```{r}
plot(x = log2(as.numeric(topGSEAft.lm[[1]][,3])), y = -log10(as.numeric(topGSEAft.lm[[2]][,5])), xlab ="log2(OR)", ylab = "-log10(adjP)", main = "OR and adjP for each enriched pathway", pch = 21, bg = "red")
```


# Session information

```{r sessionInfo, echo=T}
sessionInfo()
```

# References

